addons/survey/controllers/main.py

```diff
@@ -548,19 +548,27 @@ def _prepare_result_dict(self, survey, current_filters=None):
         current_filters = current_filters if current_filters else []
         Survey = request.env['survey.survey']
         result = {'page_ids': []}
-        for page in survey.page_ids:
-            page_dict = {'page': page, 'question_ids': []}
-            for question in page.question_ids:
+        page_dict = {}
+        for question in survey.question_and_page_ids:
+            if question.is_page:
+                # if we are on a new section, we append the previous one
+                if page_dict:
+                    result['page_ids'].append(page_dict)
+                page_dict = {'page': question, 'question_ids': []}
+            else:
+                # in case question has no section, we fill in with an empty one
+                if not question.page_id and not page_dict:
+                    page_dict = {'page': request.env['survey.question'], 'question_ids': []}
```

Comment:
Even if your code is correct, it looks to me a bit complicated to understand (use object from previous iteration before re-initialize it..).
I know what follows is a bit more verbose but I propose this for better readability.
What do you think ?

```
# First append questions without page
questions_without_page = [self._prepare_question_values(question) for question in survey.question_ids if not question.page_id]
if questions_without_page:
    result['page_ids'].append({'page': request.env['survey.question'], 'question_ids': questions_without_page})

# Then, questions in sections
for page in survey.page_ids:
    page_dict = {'page': page, 'question_ids': [self._prepare_question_values(question) for question in page.question_ids]}
    result['page_ids'].append(page_dict)
```
And
```
def _prepare_question_values(self, question, current_filters):
    Survey = request.env['survey.survey']
    return {
        'question': question,
        'input_summary': Survey.get_input_summary(question, current_filters),
        'prepare_result': Survey.prepare_result(question, current_filters),
        'graph_data': self._get_graph_data(question, current_filters),
    }
```

Reply:
@dbeguin 

Thank you.
I agree with you, it is better readable that way (and not that verbose either after all)

I have adapted (and tested), can you review?